### PR TITLE
Introduce arrow functions, rest params, and string templates to Ember Metal

### PR DESCRIPTION
**Pull request by chadhietala**
This introduces ES6 arrow functions, rest params, and string templates to Ember Metal.

### CODE REVIEW

**Comment by kimroen** [comment]
I don't think it's appropriate to use the arrow function most of the places you have done so here.

**Comment by chadhietala** [comment]
The arrow functions are shorthand and preserve lexical `this`, I don't really see the misusage.

**Comment by kimroen** [comment]
I could obviously be wrong, but the way I read this now:

```js
AliasedProperty.prototype.readOnly = function() {
  this.set = AliasedProperty_readOnlySet;
  return this;
}
```

vs

```js
AliasedProperty.prototype.readOnly = () => {
  this.set = AliasedProperty_readOnlySet;
  return this;
}
```

Before, `this` was the class prototype, but now, `this` is the surrounding context.

**Comment by stefanpenner** [comment]
> I could obviously be wrong, but the way I read this now:

you are correct, this is in error.

**Comment by stefanpenner** [comment]
@rwjblue thoughts on restargs? They are somewhat dubious in some cases still, and the `arguments` object scares me.

File: packages/ember-metal/lib/mixin.js [hunk 1]
```diff
@@ -317,11 +317,11 @@ function mergeMixins(mixins, m, descs, values, base, keys) {
       if (currentMixin._without) { a_forEach.call(currentMixin._without, removeKeys); }
     }
   }
-}
+};
 
 var IS_BINDING = /^.+Binding$/;
 
-function detectBinding(obj, key, value, m) {
+var detectBinding = (obj, key, value, m) => {
```
Comment (stefanpenner) [inline packages/ember-metal/lib/mixin.js]: i would prefer this to remain as a `function` declaration, as it has different hoisting rules.

**Comment by chadhietala** [comment]
@kimroen Yes thats a bug.

**Comment by chadhietala** [comment]
@stefanpenner Yes, Babel's de-sugaring turns rest args to a slice on `arguments` which is probably going to raise a deop, however this should not add any more deops as I did not replace areas where we were allocing an array via `var args = new Array(arguments.length)` and then setting the arguments into the new array.

**Comment by stefanpenner** [comment]
> Babel's de-sugaring turns rest args to a slice on arguments

no, it does the new array alloc: http://babeljs.io/repl/#?experimental=true&playground=false&evaluate=true&loose=false&spec=false&code=function%20asdf(b%2C%20...foo)%20%7B%0A%20%20%20var%20a%20%3D%20foo%5B1%5D%3B%0A%7D

**Comment by chadhietala** [comment]
@stefanpenner ah, it used to slice arguments like 2 weeks ago.

File: packages/ember-metal/lib/map.js [hunk 2]
```diff
@@ -182,7 +182,7 @@ OrderedSet.prototype = {
     @param {Function} fn
     @param self
   */
-  forEach: function(fn /*, thisArg*/) {
+  forEach: function(fn, ...thisArg) {
```
Comment (stefanpenner) [inline packages/ember-metal/lib/map.js]: @chadhietala so in this case, we are actually now allocating a new array, but before this PR its was merely a single index lookup. This (unless babel improves this) would be somewhat slower, but many extra bytes.

http://babeljs.io/repl/#?experimental=true&playground=false&evaluate=true&loose=false&spec=false&code=function%20before(b%2F*...foo*%2F)%20%7B%0A%20%20var%20a%20%3D%20arguments%5B1%5D%3B%0A%7D%0A%20%20%0Afunction%20after(b%2C%20...foo)%20%7B%0A%20%20%20var%20a%20%3D%20foo%5B1%5D%3B%0A%7D

File: packages/ember-metal/lib/mixin.js [hunk 3]
```diff
@@ -859,7 +859,7 @@ export function observer() {
   @param {Function} func
   @return func
 */
-export function immediateObserver() {
+export var immediateObserver = () => {
```
Comment (stefanpenner) [inline packages/ember-metal/lib/mixin.js]: i think the function declaration is more correct here.

File: packages/ember-metal/lib/observer.js [hunk 4]
```diff
@@ -17,13 +17,13 @@ import {
...
 var BEFORE_OBSERVERS = ':before';
 
-function changeEvent(keyName) {
+var changeEvent = (keyName) => {
   return keyName + AFTER_OBSERVERS;
-}
+};
 
-function beforeEvent(keyName) {
+var beforeEvent = (keyName) => {
```
Comment (stefanpenner) [inline packages/ember-metal/lib/observer.js]: and here

File: packages/ember-metal/lib/observer.js [hunk 5]
```diff
@@ -17,13 +17,13 @@ import {
 var AFTER_OBSERVERS = ':change';
 var BEFORE_OBSERVERS = ':before';
 
-function changeEvent(keyName) {
+var changeEvent = (keyName) => {
```
Comment (stefanpenner) [inline packages/ember-metal/lib/observer.js]: and here

File: packages/ember-metal/lib/observer.js [hunk 6]
```diff
@@ -40,7 +40,7 @@ export function addObserver(obj, _path, target, method) {
   return this;
 }
 
-export function observersFor(obj, path) {
+export var observersFor = (obj, path) => {
```
Comment (stefanpenner) [inline packages/ember-metal/lib/observer.js]: and here

File: packages/ember-metal/lib/observer.js [hunk 7]
```diff
@@ -78,27 +78,27 @@ export function addBeforeObserver(obj, path, target, method) {
...
 // This should only be used by the target of the observer
 // while it is setting the observed path.
-export function _suspendBeforeObserver(obj, path, target, method, callback) {
+export var _suspendBeforeObserver = (obj, path, target, method, callback) => {
   return suspendListener(obj, beforeEvent(path), target, method, callback);
-}
+};
 
-export function _suspendObserver(obj, path, target, method, callback) {
+export var _suspendObserver = (obj, path, target, method, callback) => {
```
Comment (stefanpenner) [inline packages/ember-metal/lib/observer.js]: and here

File: packages/ember-metal/lib/observer.js [hunk 8]
```diff
@@ -78,27 +78,27 @@ export function addBeforeObserver(obj, path, target, method) {
...
 
-export function _suspendBeforeObservers(obj, paths, target, method, callback) {
+export var _suspendBeforeObservers = (obj, paths, target, method, callback) => {
   var events = map.call(paths, beforeEvent);
   return suspendListeners(obj, events, target, method, callback);
-}
+};
 
-export function _suspendObservers(obj, paths, target, method, callback) {
+export var _suspendObservers = (obj, paths, target, method, callback) => {
```
Comment (stefanpenner) [inline packages/ember-metal/lib/observer.js]: etc..

File: packages/ember-metal/lib/path_cache.js [hunk 9]
```diff
@@ -4,12 +4,12 @@ var IS_GLOBAL      = /^([A-Z$]|([0-9][A-Z$]))/;
 var IS_GLOBAL_PATH = /^([A-Z$]|([0-9][A-Z$])).*[\.]/;
 var HAS_THIS       = 'this.';
 
-var isGlobalCache       = new Cache(1000, function(key) { return IS_GLOBAL.test(key);          });
-var isGlobalPathCache   = new Cache(1000, function(key) { return IS_GLOBAL_PATH.test(key);     });
-var hasThisCache        = new Cache(1000, function(key) { return key.lastIndexOf(HAS_THIS, 0) === 0; });
-var firstDotIndexCache  = new Cache(1000, function(key) { return key.indexOf('.');             });
+var isGlobalCache       = new Cache(1000, (key) => { return IS_GLOBAL.test(key);          });
```
Comment (stefanpenner) [inline packages/ember-metal/lib/path_cache.js]: great usage, but the `{ }` are not needed

**Comment by stefanpenner** [comment]
we need to decide on some style guides here.

my thoughts:

bad:
`export var foo = function() { }`

good:
`export function foo() { }`

exception:

```js
var foo = something ? function() { a } : function() { b; };
export var foo = foo;
```

reasoning:

hoisting, and clarity

**Comment by chadhietala** [comment]
@stefanpenner Agreed need styleguide.  I will need to go back and change back all the `var`s, I already addressed the ones you pointed out.

**Comment by stefanpenner** [comment]
@knownasilya you are our style guide guru, thoughts?

**Comment by knownasilya** [comment]
@stefanpenner we are on the same page

File: packages/ember-metal/lib/computed_macros.js [hunk 10]
```diff
@@ -21,10 +21,10 @@ function getProperties(self, propertyNames) {
 }
 
 function generateComputedWithProperties(macro) {
-  return function() {
+  return () => {
     var properties = a_slice.call(arguments);
 
-    var computedFunc = computed(function() {
+    var computedFunc = computed(() => {
```
Comment (kimroen) [inline packages/ember-metal/lib/computed_macros.js]: We should not use arrow function here - we don't want to change `this` in computed macros (it refers to the object the computed is used in). Many examples of this in this file - I won't comment on all.

File: packages/ember-metal/lib/deprecate_property.js [hunk 11]
```diff
@@ -23,7 +23,7 @@ import { set } from "ember-metal/property_set";
 
 export function deprecateProperty(object, deprecatedKey, newKey) {
   function deprecate() {
-    Ember.deprecate('Usage of `' + deprecatedKey + '` is deprecated, use `' + newKey + '` instead.');
+    Ember.deprecate(`Usage of "${deprecatedKey}" is deprecated, use "${newKey}" instead.`);
```
Comment (kimroen) [inline packages/ember-metal/lib/deprecate_property.js]: This is interesting. I feel like they wrapped `deprecatedKey` in ` on purpose here (it's a keyword) - we should probably keep doing that.

```js
Ember.deprecate(`Usage of \`${deprecatedKey}\` is deprecated, use \`${newKey}\` instead.`);
```

File: packages/ember-metal/lib/injected_property.js [hunk 12]
```diff
@@ -27,9 +27,9 @@ function injectedPropertyGet(keyName) {
   var possibleDesc = this[keyName];
   var desc = (possibleDesc !== null && typeof possibleDesc === 'object' && possibleDesc.isDescriptor) ? possibleDesc : undefined;
 
-  Ember.assert("Attempting to lookup an injected property on an object " +
-               "without a container, ensure that the object was " +
-               "instantiated via a container.", this.container);
+  Ember.assert(`Attempting to lookup an injected property on an object
+                without a container, ensure that the object was
+                instantiated via a container.`, this.container);
```
Comment (kimroen) [inline packages/ember-metal/lib/injected_property.js]: I'm not sure we want the whitespace here, but maybe we do?

File: packages/ember-metal/lib/deprecate_property.js (hunk 11 above)
Comment (knownasilya) [inline packages/ember-metal/lib/deprecate_property.js]: Agreed, it lets the developer know that it's an exact attribute and that's what they should look for to fix this.

File: packages/ember-metal/lib/mixin.js [hunk 13]
```diff
@@ -221,9 +221,9 @@ function applyConcatenatedProperties(obj, key, value, values) {
...
     return makeArray(value);
   }
-}
+};
 
-function applyMergedProperties(obj, key, value, values) {
+var applyMergedProperties = (obj, key, value, values) => {
   var baseValue = values[key] || obj[key];
 
   Ember.assert("You passed in `" + JSON.stringify(value) + "` as the value for `" + key +
```
Comment (kimroen) [inline packages/ember-metal/lib/mixin.js]: Missed template string?

File: packages/ember-metal/lib/mixin.js [hunk 14]
```diff
@@ -590,7 +590,7 @@ Ember.anyUnprocessedMixins = false;
   @static
   @param arguments*
 */
-Mixin.create = function() {
+Mixin.create = () => {
   // ES6TODO: this relies on a global state?
   Ember.anyUnprocessedMixins = true;
   var M = this;
```
Comment (kimroen) [inline packages/ember-metal/lib/mixin.js]: `this` reference will be the the wrong scope, but I'm curious about the `ES6TODO` comment here.

File: packages/ember-metal/lib/properties.js [hunk 15]
```diff
@@ -32,13 +32,13 @@ export function Descriptor() {
 // DEFINING PROPERTIES API
 //
 
-export function MANDATORY_SETTER_FUNCTION(name) {
+export var MANDATORY_SETTER_FUNCTION = (name) => {
   return function SETTER_FUNCTION(value) {
-    Ember.assert("You must use Ember.set() to set the `" + name + "` property (of " + this + ") to `" + value + "`.", false);
+    Ember.assert(`You must use Ember.set() to set the '${name}' property (of ${this}) to '${value}'.`, false);
```
Comment (kimroen) [inline packages/ember-metal/lib/properties.js]: Same here about ` around the attributes.

File: packages/ember-metal/lib/run_loop.js [hunk 16]
```diff
@@ -159,10 +159,9 @@ run.join = function() {
   @return {Function} returns a new function that will always have a particular context
   @since 1.4.0
 */
-run.bind = function(target, method /* args */) {
-  var args = slice.call(arguments);
-  return function() {
-    return run.join.apply(run, args.concat(slice.call(arguments)));
+run.bind = (target, method, ...args) => {
+  return (...argz) => {
```
Comment (kimroen) [inline packages/ember-metal/lib/run_loop.js]: typo `argz`

Comment (kimroen) [inline packages/ember-metal/lib/run_loop.js]: Same below. (or maybe `args` is the typo? ;) )

File: packages/ember-metal/lib/run_loop.js [hunk 17]
```diff
@@ -625,15 +623,15 @@ run.debounce = function() {
...
+run.throttle = () => {
   return backburner.throttle.apply(backburner, arguments);
 };
 
 // Make sure it's not an autorun during testing
 function checkAutoRun() {
   if (!run.currentRunLoop) {
-    Ember.assert("You have turned on testing mode, which disabled the run-loop's autorun." +
-                 " You will need to wrap any code with asynchronous side-effects in a run", !Ember.testing);
+    Ember.assert(`You have turned on testing mode, which disabled the run-loop\'s autorun.
```
Comment (kimroen) [inline packages/ember-metal/lib/run_loop.js]: This escaping is unnecessary (it doesn't do anything wrong, but it's not needed).

File: packages/ember-metal/lib/streams/stream.js [hunk 18]
```diff
@@ -229,11 +229,11 @@ Stream.prototype = {
     }
   },
 
-  valueFn: function() {
+  valueFn: () => {
     throw new Error("Stream error: valueFn not implemented");
   },
 
-  setValue: function() {
+  setValue: () => {
```
Comment (kimroen) [inline packages/ember-metal/lib/streams/stream.js]: All of these can probably be in the new syntax, like you've done elsewhere:

```js
{
  setValue() {
    ...
  }
}
```

File: packages/ember-metal/lib/utils.js [hunk 19]
```diff
@@ -676,7 +676,7 @@ var needsFinallyFix = (function() {
 
 var tryFinally;
 if (needsFinallyFix) {
-  tryFinally = function(tryable, finalizer, binding) {
+  tryFinally = (tryable, finalizer, binding) => {
     var result, finalResult, finalError;
 
     binding = binding || this;
```
Comment (kimroen) [inline packages/ember-metal/lib/utils.js]: Again, we're changing the context for `this` here, which we probably don't want.

File: packages/ember-metal/lib/utils.js [hunk 20]
```diff
@@ -696,7 +696,7 @@ if (needsFinallyFix) {
     return (finalResult === undefined) ? result : finalResult;
   };
 } else {
-  tryFinally = function(tryable, finalizer, binding) {
+  tryFinally = (tryable, finalizer, binding) => {
     var result, finalResult;
 
     binding = binding || this;
```
Comment (kimroen) [inline packages/ember-metal/lib/utils.js]: And here.

File: packages/ember-metal/lib/utils.js [hunk 21]
```diff
@@ -774,7 +774,7 @@ if (needsFinallyFix) {
     return (finalResult === undefined) ? result : finalResult;
   };
 } else {
-  tryCatchFinally = function(tryable, catchable, finalizer, binding) {
+  tryCatchFinally = (tryable, catchable, finalizer, binding) => {
     var result, finalResult;
 
     binding = binding || this;
```
Comment (kimroen) [inline packages/ember-metal/lib/utils.js]: And here.

File: packages/ember-metal/lib/deprecate_property.js (hunk 11 above)
Comment (kimroen) [inline packages/ember-metal/lib/deprecate_property.js]: The consistency on doing this throughout the code isn't that great right now, but worth not slipping at least :)

File: packages/ember-metal/lib/mixin.js [hunk 22]
```diff
@@ -252,9 +252,9 @@ function applyMergedProperties(obj, key, value, values) {
   }
 
   return newBase;
-}
+};
 
-function addNormalizedProperty(base, key, value, meta, descs, values, concats, mergings) {
+var addNormalizedProperty = (base, key, value, meta, descs, values, concats, mergings) => {
```
Comment (knownasilya) [inline packages/ember-metal/lib/mixin.js]: On thing I'd consider from a style-guide sense, is to keep the named function form if possible. It's much cleaner, and easy to see the functions.

Arrow functions should be used when functions are passed to functions, like:

```js
items.forEach(item => parent.remove(item));
```

Arrow functions could be a way into sloppy programming where the context is passed around everywhere. We want to only preserve the context when necessary.

Comment (kimroen) [inline packages/ember-metal/lib/mixin.js]: I agree.

**Comment by sebmck** [comment]
@chadhietala  

> ah, it used to slice arguments like 2 weeks ago.

No it didn't. It's been like that since v2.6.0..

File: packages/ember-metal/lib/streams/stream.js (hunk 18 above)
Comment (chadhietala) [inline packages/ember-metal/lib/streams/stream.js]: We haven't whitelisted this syntax yet.

File: packages/ember-metal/lib/path_cache.js [hunk 23]
```diff
@@ -4,10 +4,10 @@ var IS_GLOBAL      = /^([A-Z$]|([0-9][A-Z$]))/;
...
 var HAS_THIS       = 'this.';
 
-var isGlobalCache       = new Cache(1000, function(key) { return IS_GLOBAL.test(key);          });
-var isGlobalPathCache   = new Cache(1000, function(key) { return IS_GLOBAL_PATH.test(key);     });
-var hasThisCache        = new Cache(1000, function(key) { return key.lastIndexOf(HAS_THIS, 0) === 0; });
-var firstDotIndexCache  = new Cache(1000, function(key) { return key.indexOf('.');             });
+var isGlobalCache       = new Cache(1000, (key) => { IS_GLOBAL.test(key); });
+var isGlobalPathCache   = new Cache(1000, (key) => { IS_GLOBAL_PATH.test(key); });
+var hasThisCache        = new Cache(1000, (key) => { return key.lastIndexOf(HAS_THIS, 0) === 0; });
+var firstDotIndexCache  = new Cache(1000, (key) => { key.indexOf('.'); });
```
Comment (kimroen) [inline packages/ember-metal/lib/path_cache.js]: The return is only implicit when there are no curly braces:

```js
var firstDotIndexCache  = new Cache(1000, (key) => { key.indexOf('.'); } ); // Returns nothing
var firstDotIndexCache  = new Cache(1000, (key) => key.indexOf('.') ); // Returns like we want
```

File: packages/ember-metal/lib/streams/stream.js (hunk 18 above)
Comment (kimroen) [inline packages/ember-metal/lib/streams/stream.js]: My bad - I thought I saw this used elsewhere. :+1:

File: packages/ember-metal/lib/path_cache.js (hunk 23 above)
Comment (stefanpenner) [inline packages/ember-metal/lib/path_cache.js]: the parens are also not needed for single args.

Comment (knownasilya) [inline packages/ember-metal/lib/path_cache.js]: If we do `var isGlobalPathCache  = new Cache(1000, (key) => { IS_GLOBAL_PATH.test(key); } );` that should not be a single line most likely..

```js
var isGlobalPathCache  = new Cache(1000, (key) => {
  IS_GLOBAL_PATH.test(key);
});
```

Comment (kimroen) [inline packages/ember-metal/lib/path_cache.js]: > The version with no brackets will be a little confusing for some, unless they know ruby or coffeescript

...or ES6?

Sure thing, but either way it needs to return the value in this particular case, so three options:

```js
// Multiline arrow function, explicit return
var isGlobalPathCache = new Cache(1000, key => {
  return IS_GLOBAL_PATH.test(key);
});

// Single line arrow function, implicit return (no curly braces)
var isGlobalPathCache = new Cache(1000, key => IS_GLOBAL_PATH.test(key));

// Oldschool anonymous function, explicit return
var isGlobalPathCache = new Cache(1000, function(key) {
  return IS_GLOBAL_PATH.test(key);
});
```

Comment (mixonic) [inline packages/ember-metal/lib/path_cache.js]: The justification for multiple lines is the same as elsewhere: We should be able to drop a debugger in there easily. Multiple lines please.

Comment (knownasilya) [inline packages/ember-metal/lib/path_cache.js]: @kimroen or ES2015(ES6) :wink:  ..didn't notice the `return` in the original code, so my example was for functions where arrows are used but need no return.

**Comment by chadhietala** [comment]
Thanks for everyone's comments in this thread. Sorry there were so many misusage/style issues. I have updated the STYLEGUIDE.md to reflect the styles discussed in this thread.  @stefanpenner I think travis might need to be bounced.

File: STYLEGUIDE.md [hunk 24]
```diff
@@ -314,6 +325,18 @@ function fooBar(opt) {
 }
 ```
 
+#Rest Parameters
+
+Since [Babel implements](https://babeljs.io/repl/#?experimental=true&playground=true&evaluate=true&loose=false&spec=false&code=function%20foo\(...args\)%20%7B%0A%20%20%0A%7D) Rest parameters in a non-leaking matter you should us them whenever applicable.
```
Comment (diogomafra) [inline STYLEGUIDE.md]: I think you meant "you should USE them", right?

**Comment by rwjblue** [comment]
restarted travis build

**Comment by chadhietala** [comment]
@stefanpenner or @rwjblue can we merge this, starting to become a rebase monster. I can squash pending the tests passing.

**Comment by stefanpenner** [comment]
yup, thanks for the remidner.

**Comment by chadhietala** [comment]
@stefanpenner All set.

**Pull request merged**